theano/gpuarray/c_code/dnn_redux.c

```diff
@@ -97,6 +97,49 @@ int APPLY_SPECIFIC(dnn_redux)(PyGpuArrayObject *input,
       PyErr_Format(PyExc_RuntimeError, "GpuArray_reshape_inplace: %s", GpuArray_error(&(*output)->ga, err));
       return 1;
     }
+
+    if (rsz == 1 && cudnnGetVersion() <= 7004) {
+      /* We must reduce some dimensions which have all size 1.
+       * cuDNN (up to 7004) does not support this case. Let's use GpuElemwise. */
+      switch (params->red_op) {
+        // Nothing to do for following cases.
+        case CUDNN_REDUCE_TENSOR_ADD: break;
+        case CUDNN_REDUCE_TENSOR_MUL: break;
+        case CUDNN_REDUCE_TENSOR_MIN: break;
+        case CUDNN_REDUCE_TENSOR_MAX: break;
+        case CUDNN_REDUCE_TENSOR_AVG: break;
+        /* Work to do for following cases.
+        AMAX (maximum on absolute values) => apply abs(output)
+        NORM1 (addition of absolute values) => apply abs(output)
+        NORM2 (square root of sum of squares) => sqroot(output^2) => abs(output)
+        So, we must apply abs(output) for all following cases.
+        */
+        case CUDNN_REDUCE_TENSOR_AMAX:
+        case CUDNN_REDUCE_TENSOR_NORM1:
+        case CUDNN_REDUCE_TENSOR_NORM2:
+        {
+            gpuelemwise_arg arg;
+            arg.name = "out";
+            arg.typecode = (*output)->ga.typecode;
+            arg.flags = GE_READ | GE_WRITE;
+            GpuElemwise* elemwise = GpuElemwise_new(c->ctx, "", "out = (out < 0 ? -out : out)", 1, &arg, p, GE_CONVERT_F16);
+            if (!elemwise) {
+                PyErr_SetString(PyExc_RuntimeError, "Unable to create GpuElemwise for output.");
+                return 1;
+            }
+            void* args[1] = { (void*)&(*output)->ga };
+            int err = GpuElemwise_call(elemwise, args, 0);
+            GpuElemwise_free(elemwise);
```

Comment:
I think creating one each time is costly. SO I think we should keep the GpuElemwise instance in the node instance.

Reply:
To create the GpuElemewise, I need the args (which typecode may vary across op calls) and the context. These infos would be hard to get from init_code_struct section, I think.

Reply:
You can init it in the c_code() and just reuse it between Theano function call for each node.

Reply:
ok

Reply:
PS: Even if I init the GpuElemwise a first time, it depends on the output typecode, which can change from one call to another (as now we used op params and we don't explicitely check the dtypes (`check_input` is False for GpuDnnReduction)):
```C++
gpuelemwise_arg arg;
arg.name = "out";
arg.typecode = (*output)->ga.typecode; // <- HERE !
        // At next call, output typecode may be different, I think.
        //  But we need it to create the elemwise.
arg.flags = GE_READ | GE_WRITE;
GpuElemwise* elemwise = GpuElemwise_new(c->ctx, "", "out = (out < 0 ? -out : out)", 1, &arg, p, GE_CONVERT_F16);
```
Isn't it a problem ?

Reply:
The dtype is fixed for a node instance. Don't use one ptr per class, but per instance and it should work well. Don't forget to free it when we class is destroyed